File: code/py/generate_data/moment_shift.py
import numpy as np
from math import floor


def get_dense_shift_normal_moment(n, p, prop=0.2, tau=0.5):
    t0 = int(np.floor(n * tau))
    d = int(np.floor(p * prop))
    mu1 = np.zeros(p)
    sigma = np.diag(np.ones(p))
    s1 = np.random.multivariate_normal(mu1, sigma, t0)
    s2 = np.random.multivariate_normal(mu1, sigma, n-t0)
    for i in np.arange(d):
        s2[:, i] = np.random.exponential(size=n-t0) - 1

    sample = np.concatenate((s1, s2), axis=0)

    return sample


def get_sparse_shift_normal_moment(n, p, prop=0.01, tau=0.5):
    t0 = int(np.floor(n * tau))
    d = int(np.floor(p * prop))
    mu1 = np.zeros(p)
    sigma = np.diag(np.ones(p))
    s1 = np.random.multivariate_normal(mu1, sigma, t0)
    s2 = np.random.multivariate_normal(mu1, sigma, n-t0)
    for i in np.arange(d):
        s2[:, i] = np.random.exponential(size=n-t0) - 1

    sample = np.concatenate((s1, s2), axis=0)

    return sample


def get_exponential(n, p):
    s = np.zeros((n, p))
    for j in np.arange(p):
        s[:, j] = np.random.exponential(size=n)

    return s

File: code/r/generate_data/moment_shift.R
get_dense_shift_normal_moment <- function(n, p, prop = 0.2, tau = 0.5){
  t0 <- floor(n*tau)
  d <- floor(p*prop)
  s1 <- matrix(rnorm(t0*p), t0, p)
  s2 <- matrix(rnorm((n-t0)*p), (n-t0), p)
  for(j in 1:d)
    s2[, j] <- rexp(n-t0) - 1
  rbind(s1, s2)
}

get_sparse_shift_normal_moment <- function(n, p, prop = 0.01, tau = 0.5){
  t0 <- floor(n*tau)
  d <- floor(p*prop)
  s1 <- matrix(rnorm(t0*p), t0, p)
  s2 <- matrix(rnorm((n-t0)*p), (n-t0), p)
  for(j in 1:d)
    s2[, j] <- rexp(n-t0) - 1
  rbind(s1, s2)
}

get_exponential <- function(n, p){
  s <- matrix(0, n, p)
  for(j in 1:p)
    s[, j] <- rexp(n)
  s
}